### PR TITLE
All deprecations on AtomGroup and friends getters and setters.

### PR DESCRIPTION
**Pull request by dotsdl**
The new topology system for #363 will arrive in 0.16.0, and this removes
redundant getters and setters of various `Atom`, `Residue`, and
`Segment` properties. These methods predated the properties that replace
them, and this commit features deprecation warnings for each with
instructions on what to use instead.

Some `Atom` properties, such as `number`, `pos`, `serials`, have themselves been
replaced in the new scheme (`index`, `position`, `id`, respectively).
The centroid method has also been removed.

<strike>The methods `AtomGroup.translate` and `AtomGroup.rotateby` no longer
take tuples of `AtomGroup` objects as optional arguments, but only take
explicit vectors. Because `Residue`, `ResidueGroup`, `Segment`, and
`SegmentGroup` are no longer `AtomGroup`s themselves, leaving this in
creates a minor convenience at the cost of method complexity. The
documentation is correspondingly complex, and in the course of updating
the docstrings to numpy-style for #363, simplifying the method also
simplified the documenation.</strike>

Also deprecated `as_Universe` function. This has yet to be discussed on
the `issue-363` branch, but discussion should happen on what its purpose
is.

Also addressed in this commit:

- [x] In #363, `__getattr__` behavior for the Groups
have changed to a slightly different scheme. No warnings have been added
for this change to the `__getattr__` methods yet. The new scheme goes
as::

    SegmentGroups __getattr__ can select by segment name
    Segment __getattr__ can select by residue name
    ResidueGroup __getattr__ can select by residue name
    Residue __getattr__ can select by atom name
    AtomGroup __getattr__ can select by atom name

Should only require the following in `AtomGroup.__getattr__`, or
pehaps override in `ResidueGroup` and add after a `super`:

```python
if isinstance(self, ResidueGroup):
    warnings.warn("In version 0.15.0 this will select "
            "residue names, not atom names ", DeprecationWarning)
```

- [x] Because `Residue`, `ResidueGroup`, `Segment`, and
`SegmentGroup` are no longer `AtomGroup`s themselves, there are methods
such as `Segment.names` that now longer exist (one would use
`Segment.atoms.names` instead). Something like:

```python
if isinstance(self, Segment):
    warnings.warn("In version 0.15.0, use `segment.atoms.names "
                  "instead.", DeprecationWarning)
```

will be needed. And since calling `SegmentGroup.names` no longer gives a single ``numpy`` array of all atom names in each segment shoved together, but instead a list of arrays, one for each segment, this will need to be noted. A decorator will be best.

- [x] I left bonds things alone as it's not my area of expertise. Any deprecations on that front should be placed by @richardjgowers.

### CODE REVIEW

**Comment by orbeckst** [comment]
Quick comment: I would want to keep the "slot in groups of atoms for `translate` and `rotate`" because I  don't want to needlessly break old functionality.
Related note: all the new XGroups should still mostly *behave* like old `AtomGroup` (e.g. `centroid()`).

**Comment by dotsdl** [comment]
@orbeckst I will reinstate the conveniences built into these transformation methods.

File: package/MDAnalysis/core/AtomGroup.py [hunk 1]
```diff
@@ -3336,6 +3378,7 @@ def __init__(self, name, id, atoms, resnum=None):
         ##if not Residue._cache.has_key(name):
         ##    Residue._cache[name] = dict([(a.name, i) for i, a in enumerate(self._atoms)])
 
+    @deprecate(message=_FIFTEEN_DEPRECATION)
```
Comment (richardjgowers) [inline package/MDAnalysis/core/AtomGroup.py]: Can these not get moved across?

**Comment by dotsdl** [comment]
Retargeted to [0.14.1](https://github.com/MDAnalysis/mdanalysis/milestones/0.14.1%20-%20Bugfixes). Between now and BPS I have zero time to finish this up, and I think we need a bit more progress on [issue-363](https://github.com/MDAnalysis/mdanalysis/tree/issue-363) given discussion in #599 to make it clear exactly what is changing as much in advance as possible.

**Comment by khuston** [comment]
I found a potential use for `as_Universe`, though it's not very elegant. If the user wants to manually add a topology attribute, they can

    u.add_TopologyAttr(mda.core.topologyattrs.Masses(np.ones(len(u.atoms))))

And then `u.atoms.masses` works, but `u.atoms[0].mass` results in `AttributeError` because the masses are not propagated from TopologyGroup to TopologyObjects. This is fixed if instead the user does

    u.add_TopologyAttr(mda.core.topologyattrs.Masses(np.ones(len(u.atoms))))
    u = mda.core.universe.as_Universe(u)

Then both `u.atoms.masses` and `u.atoms[0].mass` work. This feels a bit hackish, and perhaps something like `as_Universe` should be performed after the user-accessible `add_TopologyAttr` is called.

Correct me if there is a better way of manually adding an attribute to the topology that gets propagated down to the individual atoms.

**Comment by richardjgowers** [comment]
@khuston hmm, adding the attribute should propogate it without your hack... I'll have a poke around

**Comment by khuston** [comment]
I'm sorry -- I'm mistaken, and the `as_Universe` is unnecessary. I must have been using the wrong version when I tried that.

**Comment by kain88-de** [comment]
@dotsdl what is still needed to get the depcreations merged?

**Comment by dotsdl** [comment]
@kain88-de I'll be addressing the comments made here and elsewhere (#599) and have this PR ready for merge by the end of the week. I've got this whole week for development, and this is among the things at the top of my list.

File: package/MDAnalysis/core/AtomGroup.py (hunk 1 above)
Comment (dotsdl) [inline package/MDAnalysis/core/AtomGroup.py]: I think these things are really fragile. They depend on particular atom names (however customary), and the implementation also seems a bit frail, depending on residue id being in order for the chain. I'd rather mark them as deprecated and then reverse the decision later if we think they can be included with confidence.

They've been included for now among the ``Atomnames`` TopologyAttr in the 363 branch (dbe6d061e35f45104317e2c3662601f19b9ea44e), but they don't work at the moment since they depend on Residue ``__getitem__`` working with atom names in a funny way. This isn't implemented yet, and I'm not sure it should be, but it is possible. See [``__getattr__`` for Residues](https://github.com/MDAnalysis/mdanalysis/blob/dbe6d061e35f45104317e2c3662601f19b9ea44e/package/MDAnalysis/core/topologyattrs.py#L278) for an example of how.

**Comment by dotsdl** [comment]
Added more deprecations; see top comment for updates. Ready for review again and hopeful for merge.

File: package/MDAnalysis/core/AtomGroup.py (hunk 1 above)
Comment (orbeckst) [inline package/MDAnalysis/core/AtomGroup.py]: "Fragile" is a very relative term. They work for pretty much anyone who cares about them, i.e. if you have protein simulations then there's a very good chance that this works because the naming is de-facto standard for the backbone atoms (termini is more varied but not as varied between the force fields and PDB that it wouldn't make sense to catch them). It's easy to take the developer's high-road and only provide functionality that "never ever will break" (haha) but that misses the point that most users are quite happy to get the work done in 90% of cases and would simply appreciate a good error message in the remaining 10%.

Comment (orbeckst) [inline package/MDAnalysis/core/AtomGroup.py]: ... so, I don't want to deprecate them.

File: package/MDAnalysis/core/AtomGroup.py [hunk 2]
```diff
@@ -457,8 +457,84 @@
 # And the return route
 _SINGULAR_PROPERTIES = {v: k for k, v in _PLURAL_PROPERTIES.items()}
 
-_FIFTEEN_DEPRECATION = "This will be removed in version 0.15.0"
+_FIFTEEN_DEPRECATION = "This will be removed in version 0.16.0"
```
Comment (kain88-de) [inline package/MDAnalysis/core/AtomGroup.py]: Can you run a global replace to `_SIXTEEN_DEPRECATION`? In case you are using emacs and projectile the short cut it `C-p r`

File: package/MDAnalysis/core/AtomGroup.py [hunk 3]
```diff
@@ -671,6 +749,7 @@ def force(self, vals):
         except (AttributeError, NoDataError):
             raise NoDataError("Timestep does not contain forces")
 
+    @deprecate(message="{}; use `position` property instead".format(_FIFTEEN_DEPRECATION))
     def centroid(self):
```
Comment (kain88-de) [inline package/MDAnalysis/core/AtomGroup.py]: what? the `centroid` isn't the same as the `position` array in my understanding.

File: package/MDAnalysis/core/AtomGroup.py [hunk 4]
```diff
@@ -685,6 +764,7 @@ def universe(self):
             return self._universe
 
     @universe.setter
+    @deprecate(message="{}; Atoms will not be able to leave their Universes.".format(_FIFTEEN_DEPRECATION))
```
Comment (kain88-de) [inline package/MDAnalysis/core/AtomGroup.py]: out of curiosity how is this triggered?

File: package/MDAnalysis/core/AtomGroup.py (hunk 1 above)
Comment (kain88-de) [inline package/MDAnalysis/core/AtomGroup.py]: Yeah rather leave a note in the doc string so people can look into it if it fails.

File: package/MDAnalysis/core/AtomGroup.py [hunk 5]
```diff
@@ -3660,14 +3803,26 @@ def __init__(self, name, residues):
         self._cls = ResidueGroup
 
     @property
+    @deprecate(message="{}; use `segid` property instead".format(_FIFTEEN_DEPRECATION))
```
Comment (kain88-de) [inline package/MDAnalysis/core/AtomGroup.py]: This conflicts with the doc string and `segid` doesn't work currently

```python
# works
u.atoms.segments[0].id
# doesn't work
u.atoms.segments[0].segid
```

File: package/MDAnalysis/core/AtomGroup.py (hunk 2 above)
Comment (kain88-de) [inline package/MDAnalysis/core/AtomGroup.py]: Or just in this file since all changes are in one file.

File: package/MDAnalysis/core/AtomGroup.py (hunk 5 above)
Comment (kain88-de) [inline package/MDAnalysis/core/AtomGroup.py]: This is important to me. When we issue deprecation messages then the suggested alternative should work.

Comment (dotsdl) [inline package/MDAnalysis/core/AtomGroup.py]: Good catch. Added a ``segid`` property to ``Segment`` to shore this up.

File: package/MDAnalysis/core/AtomGroup.py (hunk 4 above)
Comment (dotsdl) [inline package/MDAnalysis/core/AtomGroup.py]: When a user tries to set the universe an ``Atom`` is a member of, this warning will appear. If you're asking where this is used in the codebase, no idea. In the new scheme though it doesn't make sense to do this.

File: package/MDAnalysis/core/AtomGroup.py (hunk 3 above)
Comment (dotsdl) [inline package/MDAnalysis/core/AtomGroup.py]: This is for an ``Atom``, for which ``centroid`` is just its position. Unless there's a reason to keep this, I think it's unnecessary at best, confusing at worst. Happy to hear dissenting opinions, though.

File: package/MDAnalysis/core/AtomGroup.py (hunk 2 above)
Comment (dotsdl) [inline package/MDAnalysis/core/AtomGroup.py]: Yeah, I'll make the change.

File: package/MDAnalysis/core/AtomGroup.py (hunk 5 above)
Comment (kain88-de) [inline package/MDAnalysis/core/AtomGroup.py]: ideally this should be done with all deprecations.

File: package/MDAnalysis/core/AtomGroup.py (hunk 3 above)
Comment (kain88-de) [inline package/MDAnalysis/core/AtomGroup.py]: I didn't see it was for an `Atom` then yes it should be removed.

Comment (richardjgowers) [inline package/MDAnalysis/core/AtomGroup.py]: I'm not in favour of centroid per se, but I'd rather keep things focused rather than throwing in changes we'd like.

File: package/MDAnalysis/core/AtomGroup.py (hunk 4 above)
Comment (richardjgowers) [inline package/MDAnalysis/core/AtomGroup.py]: It used to be that Parser created Atoms, then the Universe assigned itself to each Atom afterwards in init

File: package/MDAnalysis/core/AtomGroup.py [hunk 6]
```diff
@@ -3660,14 +3804,35 @@ def __init__(self, name, residues):
...
+    def segid(self):
+        """Segment id (alias for :attr:`Segment.name`)"""
+        return self.name
+
+    @id.setter
+    def segid(self, x):
+        self.name = x
+
+    @property
+    @deprecate(message="{}; use `segid` property instead".format(_SIXTEEN_DEPRECATION))
```
Comment (kain88-de) [inline package/MDAnalysis/core/AtomGroup.py]: Is git messing up or are you really *adding* deprecated features here?

Comment (dotsdl) [inline package/MDAnalysis/core/AtomGroup.py]: I think git's messing up. I only added the ``segid`` property, which for the moment refers to segment ``name`` internally.

Comment (dotsdl) [inline package/MDAnalysis/core/AtomGroup.py]: Since ``id`` also uses name (see left of screen), it would be really bad if ``name`` wasn't already there.

File: package/MDAnalysis/core/AtomGroup.py (hunk 3 above)
Comment (richardjgowers) [inline package/MDAnalysis/core/AtomGroup.py]: I mean keep the PR focused, `centroid` deserves a fair trial :D

**Comment by richardjgowers** [comment]
So bonds won't work any differently in the future, apart from the fact they're now only accessible from Atom/AtomGroup.  I've added deprecations for this

**Comment by dotsdl** [comment]
Okay, so I hope nobody uses ``Atom.id`` as they currently exist because they apparently get assigned by residues. There is no corresponding mechanism in the new system (there shouldn't be; it's very unnecessary and just asking for staleness issues). 

I'm shoring up the suggestion to use ``AtomGroup.ids`` instead of ``AtomGroup.serials`` in the deprecation warning by giving the serials back if they exist, but otherwise giving back ``AtomGroup.indices``. I'm also removing the setting of ``Atom.id`` [by residues as they are built](https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/core/AtomGroup.py#L3329) because we'd like this to work the same for individual ``Atom``s as it does for ``AtomGroup``.

**Comment by dotsdl** [comment]
Alright, we should be set, then. Ready to merge if there is nothing further.

File: package/MDAnalysis/core/AtomGroup.py [hunk 7]
```diff
@@ -4671,12 +4865,14 @@ def trajectory(self, value):
         del self._trajectory  # guarantees that files are closed (?)
         self._trajectory = value
 
+    @deprecate(message=_SIXTEEN_DEPRECATION)
```
Comment (orbeckst) [inline package/MDAnalysis/core/AtomGroup.py]: Will we not need anchoring anymore?

File: package/MDAnalysis/core/AtomGroup.py [hunk 8]
```diff
@@ -4715,6 +4915,7 @@ def _matches_unpickling(self, anchor_name, n_atoms, fname, trajname):
             return False
 
 
+@deprecate(message=_SIXTEEN_DEPRECATION)
```
Comment (orbeckst) [inline package/MDAnalysis/core/AtomGroup.py]: This *might* get un-deprecated... as you said in your summary.

File: package/MDAnalysis/core/AtomGroup.py (hunk 7 above)
Comment (dotsdl) [inline package/MDAnalysis/core/AtomGroup.py]: What does it do right now? It doesn't exist as a concept in 363, so I removed it, but is it necessary given that AtomGroups no longer point to Atoms, and that Atoms no longer exist without a Universe?

File: package/MDAnalysis/core/AtomGroup.py (hunk 8 above)
Comment (dotsdl) [inline package/MDAnalysis/core/AtomGroup.py]: Yeah, it doesn't make sense at the moment in 363, so best to deprecate and then not remove it at all if it ends up being needed.

**Comment by orbeckst** [comment]
Good work @dotsdl , as far as I can tell. But I am not deeply involved in #363 so perhaps could @kain88-de or @richardjgowers take charge of the final merge? 

Should we already have a discussion with @tylerjereddy regarding new fixers for "ten2sixteen"?

File: package/MDAnalysis/core/AtomGroup.py (hunk 7 above)
Comment (orbeckst) [inline package/MDAnalysis/core/AtomGroup.py]: @mnmelo implemented it; it's required for pickling AtomGroup IIRC.

On 30 Mar, 2016, at 16:04, David Dotson wrote:

> What does it do right now? It doesn't exist as a concept in 363, so I removed it, but is it necessary given that AtomGroups no longer point to Atoms, and that Atoms no longer exist without a Universe?

--
Oliver Beckstein * orbeckst@gmx.net
skype: orbeckst  * orbeckst@gmail.com

File: package/MDAnalysis/core/AtomGroup.py [hunk 9]
```diff
@@ -457,7 +457,84 @@
 # And the return route
 _SINGULAR_PROPERTIES = {v: k for k, v in _PLURAL_PROPERTIES.items()}
 
-_FIFTEEN_DEPRECATION = "This will be removed in version 0.15.0"
+_SIXTEEN_DEPRECATION = "This will be removed in version 0.16.0"
```
Comment (kain88-de) [inline package/MDAnalysis/core/AtomGroup.py]: Looks like the deprecation warning was something older. I thought you introduced it in this PR. Did we by accident reschedule some deprecations now? Looking over the diff yes we did reschedule some deprecations.

Could you go over that again @dotsdl ?

We need a little bit more formal way to deal with deprecations and how to check that we remove functions for the targeted releases.

Comment (orbeckst) [inline package/MDAnalysis/core/AtomGroup.py]: On 31 Mar, 2016, at 00:17, kain88-de wrote:

> We need a little bit more formal way to deal with deprecations and how to check that we remove functions for the targeted releases.

Do you mean like adding a test that *fails* if a function still exists in a release?

In this particular case I have rather been advocating tests that test what we want to keep https://github.com/MDAnalysis/mdanalysis/issues/703

Comment (dotsdl) [inline package/MDAnalysis/core/AtomGroup.py]: The reason this went from "FIFTEEN" to "SIXTEEN" was because we were originally planning for the topology refactor to land in 0.15.0, with the release we're working toward now being a point release (0.14.1). Since we decided to release next as 0.15.0, this moved the topology refactor merge to 0.16.0.

Does that clarify things? I realize that it would be nice to have a more formal process for managing deprecations, but I think a major internal refactor like this is a rarity and this is about as clean as it probably gets.

Comment (kain88-de) [inline package/MDAnalysis/core/AtomGroup.py]: If this change for the other deprecations is intentional I'm fine with it. 

What I was referring that we currently clean up *a lot* and deprecate old stuff all over. So much that I forget what is supposed to be deprecated when myself. Just a list that we try to keep up to date for us devs to see what deprecations to check for when a release is happening.

**Comment by kain88-de** [comment]
Looks good to me. I've you add it to the CHANGELOG I think we are good to go.

Since all deprecations already have a replacement I think @tylerjereddy could make changes for a `ten2sixteen` script that people can already use to be ahead of the curve.

File: package/MDAnalysis/core/AtomGroup.py (hunk 9 above)
Comment (orbeckst) [inline package/MDAnalysis/core/AtomGroup.py]: On 31 Mar, 2016, at 11:04, David Dotson wrote:

> The reason this went from "FIFTEEN" to "SIXTEEN" was because we were originally planning for the topology refactor to land in 0.15.0, with the release we're working toward now being a point release (0.14.1).

Er, I think the next release WILL be 0.15.0... we already have too much new stuff (see CHANGELOG). It's just the current version tag is still 0.14.1.

--
Oliver Beckstein * orbeckst@gmx.net
skype: orbeckst  * orbeckst@gmail.com

Comment (dotsdl) [inline package/MDAnalysis/core/AtomGroup.py]: @orbeckst that's what I said. The changes were too substantial to warrant a point release. So, there's nothing to worry about with why these all went from being deprecations of things that will go missing in 0.15.0 to things that will go missing in 0.16.0. We just shifted the big #363 merge a release.

**Comment by dotsdl** [comment]
Hold up...I missed some things. For one thing, `Residue` doesn't have `resid` or `resname`, but instead using `id` and `name`. *sigh*

Adding properties for these things, and the appropriate deprecation for the alternative that should be used.

**Comment by dotsdl** [comment]
So, did a lot in the last commit (b399e183430b564af3c5da35400075ce2667556e). See log for details.

What needs doing: we get deprecation warnings on load of a `Universe` with a PSF + DCD. This is probably because the readers for these might be calling deprecated components of `MDAnalysis.core.AtomGroup.py`. Could someone that's not me try loading up a bunch of different Universes and changing the problem spots to using our "recommended" accessors/getters?

tldr: our own code shouldn't throw deprecation warnings. Only user code should.

**Comment by richardjgowers** [comment]
Ok I'm reviewing this

**Comment by tylerjereddy** [comment]
So, we want a separate `ten2sixteen` or we want to update the old one to have this new name instead of `ten2eleven`? Are the changes all summarized nicely in a single place? I guess a lot of them are mentioned or can be deduced from this PR, though end users would likely want a summary somewhere anyway.

**Comment by richardjgowers** [comment]
@dotsdl loading PSF/DCD universe doesn't throw warnings now, except.....

The anchoring stuff.  I read through it all, and there's no reason why it won't work in the new scheme.  I'm not sure how useful it is to pickle AtomGroups? but it should still work as best I can tell.  Maybe remove these deprecations and raise an Issue if we want to axe this feature?

**Comment by dotsdl** [comment]
@richardjgowers I don't think it makes any sense to pickle AtomGroups in the new scheme, since they are useless without a ``Topology``, and therefore really a ``Universe``. I think the idea of pickling AtomGroups should be abandoned in favor of making ``Universe`` pickle just fine so it can be used directly as arguments for functions using multiprocessing. Of course this means that we must be clear what pickling a ``Universe`` means (probably fully serialized Topology (#643) plus paths to trajectory(s)).

But that's another discussion. Let's de-deprecate the anchoring machinery so it stops throwing warnings from internal use. I doubt users are using this stuff directly, though I could be surprised. 

We should put a deprecation warning on AtomGroup's  `__getstate__` and `__setstate__` methods if we want to discourage pickling them.

**Comment by richardjgowers** [comment]
@tylerjereddy we started making a list [here](https://github.com/MDAnalysis/mdanalysis/wiki/Issue363-Changes) but it needs expanding a lot.  I'll try and fill it out today

**Comment by dotsdl** [comment]
Okay, after some last fixes on some omissions, I think we're finally ready. CHANGELOG updated, too. These deprecation warnings should help us out when adapting the existing tests to the new API conventions, since currently the test suite throws them everywhere.

**Comment by orbeckst** [comment]
On 2 Apr, 2016, at 09:46, Tyler Reddy wrote:

> So, we want a separate ten2sixteen or we want to update the old one to have this new name instead of ten2eleven?

Assuming that there's no harm in running "ten2sixteen" on >0.11 code,  I'd say make a new ten2sixteen and remove ten2eleven ... we just mustn't forget to update wiki etc.

**Comment by dotsdl** [comment]
Anything else to address here? Otherwise can we merge?

**Comment by kain88-de** [comment]
This results in a lot of warning messages now about deprecations.  It should be easy enough to grep for the words and replace them but I lost track what they all were. @dotsdl  can you make a list of the deprecated methods and their preferred alternative, then I can change them across the whole codebase (emacs has a function for that :wink:)

**Comment by kain88-de** [comment]
Plus I get deprecation warning when I use the preferred methods `ag.positions` triggers a warning that `get_positions` is deprecated.

**Comment by richardjgowers** [comment]
@kain88-de I'm working on this a little.  positions is because it goes through get_positions, so it needs to get redone so that get_positions goes via positions.  Ideally we should be able to run the tests with no deprecation warnings

File: package/MDAnalysis/core/AtomGroup.py [hunk 10]
```diff
@@ -1143,6 +1234,7 @@ def n_segments(self):
         return len(self.segments)
 
     @property
+    @warn_atom_property
```
Comment (kain88-de) [inline package/MDAnalysis/core/AtomGroup.py]: which one should be used instead in the future?

Comment (dotsdl) [inline package/MDAnalysis/core/AtomGroup.py]: ``*Group.atoms.indices``

Indices always refer to atom indices (zero-based, and always unique among atoms), no matter what level called from. In order to get back indices as they currently work from all levels (except AtomGroup) in the new system, one must prefix with ``.atoms``.
